### PR TITLE
Merge dev into master

### PR DESCRIPTION
**Pull request by Cecilia-Sensalari**
- Rename terms according to those used in the preprint (documentation and code lines)
- Catch missing species in configuration file fields fasta_filenames and latin_names

### CODE REVIEW

File: ksrates/fc_manipulate_trees.py [hunk 1]
```diff
@@ -426,19 +430,19 @@ def get_rates_from_current_analysis(rate_dict, correction_table, species, specie
 
 def get_rates_from_ortholog_peak_db(rate_dict, sister_node, latin_names, ortholog_db, peak_stats, missing_ortholog_data_from_database):
     """
-    It's possible that some branches in the tree can't be assigned a length proportional to relative rates based only on the data\\
-    coming from this current correction/analysis. In this case, the code tries to compute the missing relative rates on the spot\\
+    It's possible that some branches in the tree can't be assigned a length equal to branch-specific Ks contributions based only on the data\\
+    coming from this current correction/analysis. In this case, the code tries to compute the missing branch-specific Ks contributions on the spot\\
```
Comment (lohausr) [inline ksrates/fc_manipulate_trees.py:434]:
```suggestion
    coming from the current rate-adjustment analysis. In this case, the code tries to compute the missing branch-specific Ks contributions on the spot\\
```

File: ksrates/fc_manipulate_trees.py [hunk 2]
```diff
@@ -426,19 +430,19 @@ def get_rates_from_current_analysis(rate_dict, correction_table, species, specie
...
     with the relative rate test formulas by looking for the missing ortholog data in the ortholog peak database: \\
-    in fact, other corrections based on other species of interest may have already provided such missing ortholog peaks needed for the RRT formulas,\\
+    in fact, other corrections based on other focal species may have already provided such missing ortholog peaks needed for the RRT formulas,\\
```
Comment (lohausr) [inline ksrates/fc_manipulate_trees.py:436]:
```suggestion
    in fact, other rate-adjustments based on other focal species may have already provided such missing ortholog peaks needed for the RRT formulas,\\
```

File: ksrates/fc_manipulate_trees.py [hunk 3]
```diff
@@ -464,14 +468,15 @@ def get_rates_from_ortholog_peak_db(rate_dict, sister_node, latin_names, ortholo
...
+                    # If not, warn the user that it has to be computed separately to have all branch length
+                    # equal to branch-specific Ks contributions
                     try:
                         __ = ortholog_db.at[latinSister1_latinSister2, 'Ortholog_Mode']
                     except Exception:
                         missing_ortholog_data_from_database = True
 
                     # Check if there are ortholog data in database to use a species as an outgroup for the two leaves
-                    # There should always been at least the species of interest of the current analysis, except if deleted from DB for some reasons
+                    # There should always been at least the focal species of the current analysis, except if deleted from DB for some reasons
```
Comment (lohausr) [inline ksrates/fc_manipulate_trees.py:479]:
```suggestion
                    # There should always be at least the focal species of the current analysis, except if deleted from DB for some reasons
```

File: ksrates/fc_manipulate_trees.py [hunk 4]
```diff
@@ -649,11 +654,11 @@ def plotting_tree(species, latin_names, original_tree, correction_table, consens
             draw_branch_length_label(divergence_node, known_distance=False)
             unknown_branch_len_style(divergence_node)
 
-    if ortholog_db.empty: # relative rates can be get only from correction_tables
-        logging.info("Getting relative rates from correction table data")
-    else: # if the ortholog DB is available, we can try to compute the rates from there too
-        logging.info("Getting relative rates from correction table data")
-        logging.info("Computing relative rates from ortholog peak data in database by applying principles of the relative rate test")
+    if ortholog_db.empty: # branch-specific Ks contributions can be get only from adjustment_tables
```
Comment (lohausr) [inline ksrates/fc_manipulate_trees.py:657]:
```suggestion
    if ortholog_db.empty: # branch-specific Ks contributions can be obtained only from adjustment_tables
```

File: ksrates/setup_correction.py [hunk 5]
```diff
@@ -19,13 +19,18 @@ def setup_correction(config_file, nextflow_flag):
...
+    trigger_exit = False # will trigger exit if FASTA or GFF files are missing or empty
     fasta_dict = config.get_fasta_dict()
     gff_dict = config.get_gff_dict(warn_empty_dict=False)
-    latin_names = config.get_latin_names()
-    if latin_names == {}:
-        logging.error("Exiting")
-        sys.exit(1)
+    # If a GFF is provided, check existence and content
+    if species_of_interest in gff_dict:
+        gff = config.get_gff_name(gff_dict, species_of_interest)
```
Comment (lohausr) [inline ksrates/setup_correction.py:30]: Why a dict? It can be only one file.

File: ksrates/setup_correction.py [hunk 6]
```diff
@@ -37,57 +42,66 @@ def setup_correction(config_file, nextflow_flag):
     # Checking if the IDs in FASTA (and GFF) files are likely to raise an error when using the PAML package.
     # For example, paml 4.4 gives "Node" Keyerror if the len(IDs) > 50, and/or there are special characters,
     # and/or there are two spaces in a row.
-    logging.info(f"Checking if sequence IDs in input files are compatible with wgd pipeline...")
-    for species in fasta_dict.keys():
+    logging.info(f"Checking input file existence and sequence IDs compatibility with wgd pipeline...")
```
Comment (lohausr) [inline ksrates/setup_correction.py:45]:
```suggestion
    logging.info(f"Checking if sequence data files exist and if sequence IDs are compatible with wgd pipeline...")
```

File: ksrates/setup_correction.py [hunk 7]
```diff
@@ -37,57 +42,66 @@ def setup_correction(config_file, nextflow_flag):
...
 
     # 1) FINDING SISTER AND OUTGROUP SPECIES FOR EACH NODE OF INTEREST
 
     logging.info("")
-    logging.info(f"Extracting ortholog trios for rate-correction [ortholog_trios_{species_of_interest}.tsv]")
+    logging.info(f"Extracting ortholog trios for rate-adjustment [ortholog_trios_{species_of_interest}.tsv]")
 
     if isinstance(max_num_outspecies, int):
-        logging.info(f"- Each divergent species pair will be corrected with at maximum {max_num_outspecies} "
+        logging.info(f"- Each divergent species pair will be adjusted with at maximum {max_num_outspecies} "
```
Comment (lohausr) [inline ksrates/setup_correction.py:81]:
```suggestion
        logging.info(f"- Each divergent species pair will be rate-adjusted with at maximum {max_num_outspecies} "
```

File: ksrates/setup_correction.py [hunk 8]
```diff
@@ -37,57 +42,66 @@ def setup_correction(config_file, nextflow_flag):
...
                      f"trios by using the closest outspecies")
         logging.info(f"  (as required in configuration file field 'maximum_number_outspecies')")
     else:
-        logging.info(f"- Each divergent species pair will be corrected by using all the possible outspecies "
+        logging.info(f"- Each divergent species pair will be adjusted by using all the possible outspecies "
```
Comment (lohausr) [inline ksrates/setup_correction.py:85]:
```suggestion
        logging.info(f"- Each divergent species pair will be rate-adjusted by using all the possible outspecies "
```

File: ksrates/fc_configfile.py [hunk 9]
```diff
@@ -98,13 +98,13 @@ def _get_clean_dict_stringent(self, dict_like_string, parameter):
 
     def get_species(self):
         """
-        Gets the config file field of the name of the species of interest.
-        The species of interest is one of the leaves of the input tree and is the one\\
+        Gets the config file field of the name of the focal species.
+        The focal species is one of the leaves of the input tree and is the one\\
         whose paralog distribution is plotted and which the correction will be relative to.
```
Comment (lohausr) [inline ksrates/fc_configfile.py:103]:
```suggestion
        whose paralog distribution is plotted and which the rate-adjustment will be relative to. 
```

File: ksrates/fc_configfile.py [hunk 10]
```diff
@@ -126,19 +126,44 @@ def get_newick_tree(self):
             logging.error("Unrecognized format for parameter newick_tree in configuration file (for example, parentheses do not match)")
             sys.exit(1)
 
+    def check_complete_dictionary(self, dictionary):
```
Comment (lohausr) [inline ksrates/fc_configfile.py:129]: If this function is specific to `latin_names` name it that way. If not document it in a more general way without referring to `latin_names`

File: ksrates/fc_configfile.py [hunk 11]
```diff
@@ -126,19 +126,44 @@ def get_newick_tree(self):
...
+        missing_species = list(set.difference(set(all_leaves), set(dictionary.keys())))
+        if len(missing_species) != 0:
+            if len(missing_species) == 1:
+                logging.error(f"The following species is missing from [latin_names] configuration file field:")
+            else:
+                logging.error(f"The following species are missing from [latin_names] configuration file field:")
+            for missing_name in missing_species:
+                logging.error(f" - {missing_name}")
+            
+            logging.error(f"Please add the missing information and rerun the analysis.")
```
Comment (lohausr) [inline ksrates/fc_configfile.py:148]:
```suggestion
            logging.error(f"Please add the missing information and restart the analysis.")
```

File: ksrates/fc_configfile.py [hunk 12]
```diff
@@ -126,19 +126,44 @@ def get_newick_tree(self):
...
+
+        :param dictionary: the dictionary coming from latin_names
+        """
+        all_leaves = []
+        for leaf in self.get_newick_tree().get_leaves():
+            all_leaves.append(leaf.name)
+        missing_species = list(set.difference(set(all_leaves), set(dictionary.keys())))
+        if len(missing_species) != 0:
+            if len(missing_species) == 1:
+                logging.error(f"The following species is missing from [latin_names] configuration file field:")
```
Comment (lohausr) [inline ksrates/fc_configfile.py:142]:
```suggestion
                logging.error(f"The following species is missing from the [latin_names] configuration file field:")
```

File: ksrates/fc_configfile.py [hunk 13]
```diff
@@ -126,19 +126,44 @@ def get_newick_tree(self):
...
+            else:
+                logging.error(f"The following species are missing from [latin_names] configuration file field:")
```
Comment (lohausr) [inline ksrates/fc_configfile.py:144]:
```suggestion
                logging.error(f"The following species are missing from the [latin_names] configuration file field:")
```

File: ksrates/fc_configfile.py [hunk 14]
```diff
@@ -126,19 +126,44 @@ def get_newick_tree(self):
...
 
         :return latin_names_dict: python dictionary
         """
         latin_names = self.config.get("SPECIES", "latin_names")
         if latin_names != "":
             latin_names_dict = self._get_clean_dict_stringent(latin_names, "latin_names")
         else:
-            logging.warning("Latin names field in configuration file is empty.")
-            latin_names_dict = {}
+            logging.error("Configuration file field [latin_names] is empty, please fill in and rerun.")
```
Comment (lohausr) [inline ksrates/fc_configfile.py:162]:
```suggestion
            logging.error("Configuration file field [latin_names] is empty, please fill in and restart the analysis.")
```

File: ksrates/fc_configfile.py [hunk 15]
```diff
@@ -173,7 +198,7 @@ def get_fasta_dict(self):
         if fasta_names_string != "":
             fasta_names_dict = self._get_clean_dict(fasta_names_string, "FASTA file")
         else:
-            logging.warning("Empty string in FASTA filename field in configuration file.")
+            logging.warning("Configuration file field [fasta_filenames] is empty")
```
Comment (lohausr) [inline ksrates/fc_configfile.py:201]:
```suggestion
            logging.warning("Configuration file field [fasta_filenames] is empty.")
```

File: ksrates/fc_configfile.py [hunk 16]
```diff
@@ -214,21 +239,21 @@ def get_gff_dict(self, warn_empty_dict=True):
 
     def get_gff_name(self, gff_dict, species):
         """
-        Gets the path to the GFF file of the species of interest from the dictionary.
+        Gets the path to the GFF file of the focal species from the dictionary.
         If the value is an empty string, it applies the fallback filename "species + .gff".
 
-        :param fasta_dict: Python dictionary that associates each informal species name to the path of its FASTA file 
-        :param species: the informal name of the species of interest
+        :param gff_dict: Python dictionary that associates the focal species informal name to the path of its GFF file
```
Comment (lohausr) [inline ksrates/fc_configfile.py:245]: Why is this necessary? It's always for the focal species.

File: ksrates/fc_configfile.py [hunk 17]
```diff
@@ -214,21 +239,21 @@ def get_gff_dict(self, warn_empty_dict=True):
...
 
-        :param fasta_dict: Python dictionary that associates each informal species name to the path of its FASTA file 
-        :param species: the informal name of the species of interest
+        :param gff_dict: Python dictionary that associates the focal species informal name to the path of its GFF file 
+        :param species: the informal name of the focal species
         :return: the GFF file path
         """
         if gff_dict == {}:
-            logging.warning(f"Default GFF filename applied for {species} ({species}.gff)")
+            logging.warning(f"GFF filename for {species} not found in configuration file; assuming default one ({species}.gff)")
```
Comment (lohausr) [inline ksrates/fc_configfile.py:250]:
```suggestion
            logging.warning(f"GFF filename for focal species {species} not found in configuration file; assuming default one ({species}.gff)")
```

File: ksrates/fc_configfile.py [hunk 18]
```diff
@@ -214,21 +239,21 @@ def get_gff_dict(self, warn_empty_dict=True):
...
             gff = f"{species}.gff"   # fallback name   
         else:
             if species in gff_dict and gff_dict[species] != "": # if the gff filename is an acceptable string (not empty)
                 gff = gff_dict[species]
             else:
-                logging.warning(f"GFF filename for {species} not found in configuration file; default one applied ({species}.gff)")
+                logging.warning(f"GFF filename for {species} not found in configuration file; assuming default one ({species}.gff)")
```
Comment (lohausr) [inline ksrates/fc_configfile.py:256]:
```suggestion
                logging.warning(f"GFF filename for focal species {species} not found in configuration file; assuming default one ({species}.gff)")
```

File: ksrates/fc_configfile.py [hunk 19]
```diff
@@ -303,7 +328,7 @@ def get_max_num_outspecies(self):
 
     def get_paranome(self):
         """
-        Gets the config file field specifying if the mixed distribution will show or not the whole-paranome of the species of interest.
+        Gets the config file field specifying if the mixed distribution will show or not the whole-paranome of the focal species.
```
Comment (lohausr) [inline ksrates/fc_configfile.py:331]:
```suggestion
        Gets the config file field specifying if the mixed distribution will show the whole-paranome of the focal species or not.
```

File: ksrates/fc_configfile.py [hunk 20]
```diff
@@ -318,7 +343,7 @@ def get_paranome(self):
 
     def get_colinearity(self):
         """
-        Gets the config file field specifying if the mixed distribution will show or not the anchor pairs of the species of interest.
+        Gets the config file field specifying if the mixed distribution will show or not the anchor pairs of the focal species.
```
Comment (lohausr) [inline ksrates/fc_configfile.py:346]:
```suggestion
        Gets the config file field specifying if the mixed distribution will show the anchor pairs of the focal species or not.
```

File: main.nf [hunk 21]
```diff
@@ -859,22 +859,22 @@ process drawTree {
...
-    echo "NF internal work directory for [drawTree (${task.index})] process:\n\$processDir\n" >> $logs_folder/rate_correction.log
+    echo "NF internal work directory for [drawTree (${task.index})] process:\n\$processDir\n" >> $logs_folder/rate_adjustment.log
 
-    ksrates plot-tree ${config} --nextflow >> $logs_folder/rate_correction.log 2>&1
+    ksrates plot-tree ${config} --nextflow >> $logs_folder/rate_adjustment.log 2>&1
 
     RET_CODE=\$?
     echo "[$species] Done [\${RET_CODE}]"
-    echo "[$species] log can be found in: $logs_folder/rate_correction.log"
+    echo "[$species] log can be found in: $logs_folder/rate_adjustment.log"
```
Comment (lohausr) [inline main.nf:877]: Why does logging for the tree go into `rate_adjustment.log`?

File: ksrates/plot_paralogs.py [hunk 22]
```diff
@@ -7,7 +7,7 @@
 import ksrates.fc_extract_ks_list as fc_extract_ks_list
 import ksrates.fc_check_input as fcCheck
 import ksrates.fc_configfile as fcConf
-
+from ksrates.fc_rrt_correction import _ADJUSTMENT_TABLE
```
[the remark this answers is not in the document]
Comment (Cecilia-Sensalari) [reply ksrates/plot_paralogs.py:10]: Since _MIXED_ADJUSTED_PLOT_FILENAME is a variable coming from a module that is imported in the first lines, I call it as `fcPlot._MIXED_ADJUSTED_PLOT_FILENAME` directly in the code (line 131).
Instead, since _ADJUSTMENT_TABLE comes from a module that is not imported in the first lines because not needed, I import only that variable from it at the beginning of the script.

File: ksrates/plot_tree.py [hunk 23]
```diff
@@ -4,27 +4,28 @@
 import ksrates.fc_manipulate_trees as fcTree
 import ksrates.fc_check_input as fcCheck
 import ksrates.fc_configfile as fcConf
+from ksrates.fc_rrt_correction import _ADJUSTMENT_TABLE
```
[the remark this answers is not in the document]
Comment (Cecilia-Sensalari) [reply ksrates/plot_tree.py:7]: Same comment for plot_paralogs.py: _ADJUSTMENT_TABLE is the only "constant" that I need to import because it doesn't come from a module that is actually used in plot_tree.py. This way I avoid importing the whole module fc_rrt_correction.py for just one variable.

**Review by Cecilia-Sensalari — COMMENTED** [review]
- Accepted suggested code blocks (they are in new commits or in a "batch" of commits)
- Pushed new changes based on your comments
- Labeled many conversations as resolved
- Responded to others

File: main.nf (hunk 21 above)
Comment (lohausr) [reply main.nf:877]: Alright, let's leave it like that.

File: ksrates/plot_paralogs.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (lohausr) [reply ksrates/plot_paralogs.py:10]: Alright.

File: ksrates/plot_paralogs.py [hunk 24]
```diff
@@ -66,7 +67,6 @@ def plot_paralogs_distr(config_file, correction_table_file, paralog_tsv_file, an
     consensus_peak_for_multiple_outgroups = config.get_consensus_peak_for_multiple_outgroups()
 
     # Get other parameters
-    latin_names = config.get_latin_names()
```
[the remark this answers is not in the document]
Comment (lohausr) [reply ksrates/plot_paralogs.py:69]: OK.

File: ksrates/plot_tree.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (lohausr) [reply ksrates/plot_tree.py:7]: Alright.

File: ksrates/setup_correction.py (hunk 5 above)
Comment (lohausr) [reply ksrates/setup_correction.py:30]: See an earlier comment. I don't think the analogy here warrants the usage of a dict, seems unnecessarily convoluted.

**Review by lohausr — CHANGES_REQUESTED** [review]
A few more things...

File: ksrates/fc_configfile.py (hunk 16 above)
Comment (Cecilia-Sensalari) [reply ksrates/fc_configfile.py:245]: I see what you mean, alright I'll change it :)

File: doc/source/configuration.rst [hunk 25]
```diff
@@ -187,22 +187,23 @@ This is an optional configuration file that contains several \"expert\" paramete
...
+* **plot_adjustment_arrows**: flag to toggle the plotting of rate-adjustment arrows below the adjusted mixed paralog--ortholog *K*:sub:`S` plot. These arrows start from the original unadjusted ortholog divergence *K*:sub:`S` estimate and end on the rate-adjusted estimate (options: "yes" and "no"). [Default: "no"]
+* **num_mixture_model_initializations**: number of times the EM algorithm is initialized (either for the random initialization in the exponential-lognormal mixture model or for k-means in the lognormal mixture model). [Default: 10]
+* **max_mixture_model_iterations**: maximum number of EM iterations for mixture modeling. [Default: 300]
+* **max_mixture_model_components**: maximum number of components considered during execution of the mixture models. [Default: 5]
+* **max_mixture_model_ks**: upper limit for the *K*:sub:`S` range in which the exponential-lognormal and lognormal-only mixture models are performed. [Default: 5]
+* **extra_paralogs_analyses_methods**: flag to toggle the optional analysis of the paralog *K*:sub:`S` distribution with non default mixture model methods. [Default: "no"]
```
[the remark this answers is not in the document]
Comment (Cecilia-Sensalari) [reply doc/source/configuration.rst:209]: I'll add an internal link to the Paralog analyses page of the documentation and I'll say also to check the Supplementary Materials.

File: ksrates/fc_exp_log_mixture.py [hunk 26]
```diff
@@ -67,8 +68,8 @@ def generate_best_model_figure(species_escape_whitespace, latin_species, x_max_l
   Generates the figure hosting the best mixture model
   (according to lowest BIC score).
 
-  :param species_escape_whitespace: scientific name of the species of interest with space escape
-  :param latin_species: scientific name of the species of interest
+  :param species_escape_whitespace: scientific name of the focal species with space escape
+  :param latin_species: scientific name of the focal species
```
[the remark this answers is not in the document]
Comment (lohausr) [reply ksrates/fc_exp_log_mixture.py:72]: I meant between these two parameter names, they both seem to be the scientific name. Now thinking about it, why are both needed, can't you get the white-escaped one internally from the other one?

File: ksrates/wgd_paralogs.py [hunk 27]
```diff
@@ -16,35 +16,43 @@ def wgd_paralogs(config_file, n_threads):
...
-    else:
-        fcCheck.check_IDs(species_fasta_file, latin_names[species])
+            logging.error("No GFF feature provided in configuration file. Will exit.")
+            trigger_exit = True
+
+    # Checking if FASTA file exists and if sequence IDs are compatible with wgd pipeline (paml)
+    fasta_names_dict = config.get_fasta_dict()
+    species_fasta_file = config.get_fasta_name(fasta_names_dict, species)
+    if fcCheck.check_file_nonexistent_or_empty(species_fasta_file, "FASTA file"):  # if missing/empty
+        trigger_exit = True
```
Comment (lohausr) [inline ksrates/wgd_paralogs.py:45]: I would have probably left doing the FASTA file check first before the GFF file check.
